jolt-core/src/zkvm/prover.rs

```diff
@@ -1137,69 +1099,143 @@ impl<'a, F: JoltField, PCS: StreamingCommitmentScheme<Field = F>, ProofTranscrip
         (Some(proof_val), proof_val_final)
     }
 
-    /// Stage 7: Batch opening reduction sumcheck.
-    /// Generates witness polynomials, runs the sumcheck, and stores state for Stage 8.
+    /// Stage 7: HammingWeight + ClaimReduction sumcheck (only log_k_chunk rounds).
+    /// Produces `DoryOpeningState` for Stage 8.
     #[tracing::instrument(skip_all)]
     fn prove_stage7(&mut self) -> SumcheckInstanceProof<F, ProofTranscript> {
-        tracing::info!("Stage 7 proving (batch opening reduction sumcheck)");
-
-        let _guard = DoryGlobals::initialize(self.one_hot_params.k_chunk, self.padded_trace_len);
+        tracing::info!("Stage 7 proving (HammingWeight claim reduction)");
 
-        let polynomials_map = CommittedPolynomial::generate_witness_batch(
-            &all_committed_polynomials(&self.one_hot_params),
-            self.preprocessing,
+        // Create params and prover for HammingWeightClaimReduction
+        // (r_cycle and r_addr_bool are extracted from Booleanity opening internally)
+        let hw_params = HammingWeightClaimReductionParams::new(
+            &self.one_hot_params,
+            &self.opening_accumulator,
+            &mut self.transcript,
+        );
+        let mut hw_prover = HammingWeightClaimReductionProver::initialize(
+            hw_params,
             &self.trace,
+            self.preprocessing,
             &self.one_hot_params,
         );
 
         #[cfg(feature = "allocative")]
-        print_data_structure_heap_usage("Committed polynomials map", &polynomials_map);
+        print_data_structure_heap_usage("HammingWeightClaimReductionProver", &hw_prover);
 
-        // Prepare sumcheck
-        self.opening_accumulator
-            .prepare_for_sumcheck(&polynomials_map);
+        // 3. Run sumcheck (only log_k_chunk rounds!)
+        let instances: Vec<&mut dyn SumcheckInstanceProver<F, ProofTranscript>> =
+            vec![&mut hw_prover];
+        let (sumcheck_proof, r_address_stage7) = BatchedSumcheck::prove(
+            instances,
+            &mut self.opening_accumulator,
+            &mut self.transcript,
+        );
 
-        // Run sumcheck
-        let (sumcheck_proof, r_sumcheck) = self
-            .opening_accumulator
-            .prove_batch_opening_sumcheck(&mut self.transcript);
+        // 4. Collect all claims for DoryOpeningState
```

Comment:
same comment; I think if we move the rest of this (i.e. steps 4-7) into `prove_stage8`, we can get rid of the `DoryOpeningState`